source/microbit/filesystem.c

```diff
@@ -0,0 +1,436 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include <string.h>
+#include <stdio.h>
+#include <errno.h>
+#include <sys/stat.h>
+
+#include "py/nlr.h"
+#include "py/obj.h"
+#include "py/gc.h"
+#include "py/stream.h"
+#include "filesystem.h"
+
+#define DEBUG_FILE 0
+#if DEBUG_FILE
+#define DEBUG(s) printf s
+#else
+#define DEBUG(s) (void)0
+#endif
+
+/**  How it works:
+ * The File System consists of CHUNKS_IN_FILE_SYSTEM chunks of CHUNK_SIZE each,
+ * plus one spare page which holds persistent configuration data and is used. for bulk erasing.
+ * The spare page is either the first or the last page and will be switched by a bulk erase.
+ *
+ * Each chunk consists of a one byte marker and a one byte tail
+ * The marker shows whether this chunk is the start of a file, the midst of a file
+ * (in which case it refers to the previous chunk in the file) or whether it is UNUSED
+ * (and erased) or FREED (which means it is unused, but not erased).
+ * Chunks are selected in a randomised round-robin fashion to even out wear on the flash
+ * memory as much as possible.
+ * A file consists of a linked list of chunks. The first chunk in a file contains its name
+ * as well as the end chunk and offset.
+ * Files are found by linear search of the chunks, this means that no meta-data needs to be stored
+ * outside of the file, which prevents wear hot-spots. Since there are fewer than 200 chunks,
+ * the search is fast enough.
+ *
+ * Writing to files relies on the persistent API which is high-level wrapper on top of the Nordic SDK.
+ */
+
+
+extern int8_t file_system_data[];
+
+STATIC_ASSERT((sizeof(file_chunk) == CHUNK_SIZE));
+
+// Pages are numbered from 1 to PAGES_IN_FILE_SYSTEM, as we need to reserve 0 as the FREED marker.
+static file_chunk *file_system_chunks = NULL;
+
+static inline void *first_page(void) {
+    return &file_system_data[0];
+}
+
+static inline void *last_page(void) {
+    return &(((file_chunk *)&file_system_data)[CHUNKS_IN_FILE_SYSTEM]);
+}
+
+void microbit_filesystem_init(void) {
+    file_chunk *base = &(((file_chunk *)&file_system_data)[0]);
```

Comment:
This line gives me a strict alias warning.  If you change it to `file_chunk *base = (file_chunk *)&file_system_data;` then it doesn't give the warning and is semantically the same.